fdbserver/RestoreApplier.actor.cpp

```diff
@@ -134,102 +134,253 @@ ACTOR static Future<Void> handleSendMutationVectorRequest(RestoreSendMutationVec
+					TraceEvent(SevWarn, "FastRestore_ApplyTxnError")
+					    .detail("TxnStatusFailed", progress.curTxnId)
+					    .detail("ApplierApplyToDB", self->id())
+					    .detail("UncommittedTxnId", progress.uncommittedTxnId)
+					    .detail("CurIteratorVersion", progress.curItInCurTxn->first)
+					    .detail("StartIteratorVersionInUncommittedTxn", progress.startItInUncommittedTxn->first)
+					    .detail("CurrentIndexInFailedTxn", progress.curIndexInCurTxn)
+					    .detail("StartIndexInUncommittedTxn", progress.startIndexInUncommittedTxn)
+					    .detail("NumIncludedAtomicOps", progress.numAtomicOps);
+					progress.rollback();
+					continue;
+				} else {
+					TraceEvent(SevWarn, "FastRestore_ApplyTxnError")
+					    .detail("TxnStatusSucceeded", progress.curTxnId)
+					    .detail("ApplierApplyToDB", self->id())
+					    .detail("CurIteratorVersion", progress.curItInCurTxn->first)
+					    .detail("CurrentIteratorMutations", progress.curItInCurTxn->second.size())
+					    .detail("CurrentIndexInSucceedTxn", progress.curIndexInCurTxn)
+					    .detail("NumIncludedAtomicOps", progress.numAtomicOps);
+
+					// Txn succeeded and exectue the same logic when txn succeeds
+				}
+			} else { // !lastTxnHasError: accumulate mutations in a txn
+				tr->reset();
+				tr->setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+				tr->setOption(FDBTransactionOptions::LOCK_AWARE);
+				TraceEvent("FastRestore_ApplierTxn")
+				    .detail("ApplierApplyToDB", self->id())
+				    .detail("TxnId", progress.curTxnId)
+				    .detail("StartIndexInCurrentTxn", progress.curIndexInCurTxn)
+				    .detail("CurrentIteratorMutations", progress.curItInCurTxn->second.size())
+				    .detail("Version", progress.curItInCurTxn->first);
+
+				// restoreApplierKeyFor(self->id(), curTxnId) to tell if txn succeeds at an unknown error
+				tr->set(restoreApplierKeyFor(self->id(), progress.curTxnId), restoreApplierTxnValue);
+
+				while (1) { // Loop: Accumulate mutations in a transaction
+					state MutationRef m;
+					ASSERT_WE_THINK(progress.curIndexInCurTxn < progress.curItInCurTxn->second.size());
+
+					m = progress.curItInCurTxn->second[progress.curIndexInCurTxn];
 					if (m.type >= MutationRef::Type::SetValue && m.type <= MutationRef::Type::MAX_ATOMIC_OP) {
 						typeStr = typeString[m.type];
-					}
-					else {
+					} else {
 						TraceEvent(SevError, "FastRestore").detail("InvalidMutationType", m.type);
 					}
 
+					//TraceEvent(SevDebug, "FastRestore_Debug").detail("ApplierApplyToDB", self->describeNode()).detail("Version", it->first).detail("Mutation", m.toString());
 					if (m.type == MutationRef::SetValue) {
 						tr->set(m.param1, m.param2);
 					} else if (m.type == MutationRef::ClearRange) {
 						KeyRangeRef mutationRange(m.param1, m.param2);
 						tr->clear(mutationRange);
 					} else if (isAtomicOp((MutationRef::Type)m.type)) {
 						tr->atomicOp(m.param1, m.param2, m.type);
+						progress.numAtomicOps++;
```

Comment:
I noticed this number is only used in trace events. If not very useful, you can safely remove it.

Reply:
I'd like to know how many atomic operations have been applied to DB. 
This can be a useful info for performance tuning later.